### PR TITLE
Fixed a bunch of compiler warnings and deprecated functions.

### PR DESCRIPTION
**Pull request by sfjuocekr**
Fixed a bunch of compiler warnings.
Fixed a bunch of deprecated functions.
Changed the setup script to allow builds on 21.04.

### CODE REVIEW

**Comment by raphaelcoeffic** [comment]
Just an idea, but I think we should make some CI/CD script to compile companion when changes to companion are committed, don’t you think?

**Comment by sfjuocekr** [comment]
Probably, but my experience regarding CI/CD is very limited. I'm one of those text editor programmers :(

I've been trying to get my head around QTCreator the past days.

File: companion/src/firmwares/opentx/opentxinterface.cpp [hunk 1]
```diff
@@ -178,14 +178,14 @@ bool OpenTxEepromInterface::loadFromByteArray(T & dest, const QByteArray & data)
   uint32_t fourcc = *((uint32_t*)&data.data()[0]);
   if (Boards::getFourCC(board) != fourcc) {
     if (IS_FAMILY_HORUS_OR_T16(board) && fourcc == 0x3178396F) {
-      qDebug() << QString().sprintf("%s: Deprecated fourcc used %x vs %x", getName(), fourcc, Boards::getFourCC(board));
+      qDebug() << QString().asprintf("%s: Deprecated fourcc used %x vs %x", getName(), fourcc, Boards::getFourCC(board));
```
Comment (elecpower) [inline companion/src/firmwares/opentx/opentxinterface.cpp:181]: As per QT doco, even asprintf is not recommended for new code so if updating it is recommended to use QTextStream or arg

File: companion/src/simulation/simulateduiwidget.cpp [hunk 2]
```diff
@@ -145,7 +145,7 @@ void SimulatedUIWidget::captureScreenshot()
       path = "./";
     QDir dir(path);
     if (dir.exists() && dir.isReadable()) {
-      QStringList fwid = getCurrentFirmware()->getId().split("-", QString::SkipEmptyParts);
+      QStringList fwid = getCurrentFirmware()->getId().split("-", Qt::KeepEmptyParts);
```
Comment (elecpower) [inline companion/src/simulation/simulateduiwidget.cpp:148]: Why the change from skip to keep?

File: companion/src/modeledit/setup.cpp [hunk 3]
```diff
@@ -261,7 +261,7 @@ ModulePanel::ModulePanel(QWidget * parent, ModelData & model, ModuleData & modul
   connect(ui->multiSubType, static_cast<void(QComboBox::*)(int)>(&QComboBox::currentIndexChanged), this, &ModulePanel::onSubTypeChanged);
   connect(ui->multiProtocol, static_cast<void(QComboBox::*)(int)>(&QComboBox::currentIndexChanged), this, &ModulePanel::onMultiProtocolChanged);
   connect(this, &ModulePanel::channelsRangeChanged, this, &ModulePanel::setupFailsafes);
-  connect(ui->btnGrpValueType, static_cast<void(QButtonGroup::*)(int)>(&QButtonGroup::buttonClicked), this, &ModulePanel::onFailsafesDisplayValueTypeChanged);
+  connect(ui->btnGrpValueType, static_cast<void(QButtonGroup::*)(int)>(&QButtonGroup::idClicked), this, &ModulePanel::onFailsafesDisplayValueTypeChanged);
```
Comment (elecpower) [inline companion/src/modeledit/setup.cpp:264]: Introduced at 5.15 so backwards compatibility issue with OpenTX and buttonClicked is still supported

**Comment by elecpower** [comment]
IMHO there needs to be agreement on what the build environment is otherwise devs will be chasing their tails trying to diagnose why things work or don't considering devs are using Windows, flavours of Linux and maybe even Mac.
I have no problem in devs using the environment of their choice but all PRs should be compiled and tested against the agreed environment.
There are lots of changes to the project code without adding moving platform. This is not to say the platform should stay fixed but maybe review and if necessary update after each major project release.

**Comment by pfeerick** [comment]
Agreed. Just like with the firmware - a set environment (or environments, if it needs to be natively built) - needs to be used for consistency and repeatability. Or at least version locks - i.e. QT 5.12, not just whatever happens to be the latest.

**Comment by raphaelcoeffic** [comment]
> Agreed. Just like with the firmware - a set environment (or environments, if it needs to be natively built) - needs to be used for consistency and repeatability. Or at least version locks - i.e. QT 5.12, not just whatever happens to be the latest.

Traditionally, this is done by having the CI/CD using the oldest version supported. Everything must be checked against that version. So bottom line is: we need CI/CD for Companion if we want to be able to do this properly. @elecpower is that what you had in mind?

File: companion/src/simulation/simulateduiwidget.cpp (hunk 2 above)
Comment (sfjuocekr) [reply companion/src/simulation/simulateduiwidget.cpp:148]: Copy paste error!

**Comment by sfjuocekr** [comment]
Some of these changes could be rewritten with #ifdef to check what version is used to silence the compiler on those versions.

**Comment by elecpower** [comment]
I agree with, where practical, suppressing compiler warnings but please no version #ifdef just to suppress compiler warnings. Set the build platform and live with any annoyances.

**Comment by elecpower** [comment]
> Traditionally, this is done by having the CI/CD using the oldest version supported. Everything must be checked against that version. So bottom line is: we need CI/CD for Companion if we want to be able to do this properly. @elecpower is that what you had in mind?

@raphaelcoeffic not quite that far but why not have that as longer term aim. CI tests would need to be across supported OS platforms. The real challenge are those obscure issues that arise between OS platforms eg UI display.

Anyway my comment was mainly to set the project's dev environment and all devs code to that and don't expend time writing and/or testing against anything else and then trying to shoehorn it into the project codebase. Some ideas in a collaborative project just have to wait until the time arises.

Also on the CI/CD theme need to consider recompiling libsimulators when changes to radios. As these are currently package with Companion so would need a new download process. This could potentially be added to existing Companion nightly function but as always some thought needed and consideration in the context in the use of EdgeTX Flasher.

**Comment by pfeerick** [comment]
@elecpower Since Companion is your baby ATM, I'll defer to you... I take it there are compatability issues making it so this PR needs more than re-signing with current codebase and merging?

**Comment by pfeerick** [comment]
I also agree with not using ifdefs to silence warnings - better to find a compiler flag to shut it up where possible ;)

**Comment by elecpower** [comment]
> I take it there are compatability issues making it so this PR needs more than re-signing with current codebase and merging?

Yes as per previous comments. Also, I do not get these warnings using the EdgeTX build environment so it would appear they are specific each dev's build environment.
IMHO we 'fix' those that pop up in the official build environment otherwise there will be PR after PR for individual dev issues that potentially cause issues for other dev's environments or even the official build environment.
These are annoying warnings not errors. But if you stray from the official path then be it on your head.
That being said in the next day or so I will recheck Qt version compatibility with OTX and flag any that are incompatible in addition to the one already noted.

**Comment by pfeerick** [comment]
Indeed... we can only *support* the documented/official dev setup, anything else will only lead to never-ending chaos. Once you've had a chance to review, lets give @sfjuocekr a chance to get this PR up to date and see if we can get it merged. 

To that end, I think I'm going to have to say we can't accept the changes to `tools/setup_buildenv_ubuntu20.04.sh` in this form, as ubuntu21 is *not* a supported build environment as yet - and those changes wouldn't be consistent with the script anyway, since it is specifically for ubuntu 21.04 ;) 

I would suggest capturing the script in it's current form (as it has changed somewhat in the mean time, make your changes, and re-adding it as `setup_buildenv_ubuntu21.04_unsupported.sh`, whilst adding a caution at the start and end of the run that this is *not* an officially supported build configuration, so use at your own risk. I would like to keep it rather reject outright, as it will allow for this build environment to actually get tested. :) Although I would suspect the next supported build environment would be 22.04 (next LTS). Your thoughts @raphaelcoeffic ?

**Comment by elecpower** [comment]
@pfeerick @raphaelcoeffic Since EdgeTX is bleeding edge and I agree we want contributions from all quarters, how about an unsupported resources area for scripts and How To instructions. Some really good ideas/techniques/tricks/etc may come from those and make it into the official environment?

File: companion/src/burnconfigdialog.cpp [hunk 4]
```diff
@@ -93,7 +93,7 @@ void burnConfigDialog::getSettings()
       dfuLoc =  QFileInfo("/usr/bin/dfu-util").absoluteFilePath();
 #endif
 
-    dfuArgs = g.dfuArguments().split(" ", QString::SkipEmptyParts);
+    dfuArgs = g.dfuArguments().split(" ", Qt::SkipEmptyParts);
```
Comment (elecpower) [inline companion/src/burnconfigdialog.cpp:96]: Needs Qt 5.14 - change not supported

File: companion/src/burnconfigdialog.cpp [hunk 5]
```diff
@@ -160,7 +160,7 @@ void burnConfigDialog::on_dfu_location_editingFinished()
 
 void burnConfigDialog::on_dfuArgs_editingFinished()
 {
-    dfuArgs = ui->dfuArgs->text().split(" ", QString::SkipEmptyParts);
+    dfuArgs = ui->dfuArgs->text().split(" ", Qt::SkipEmptyParts);
```
Comment (elecpower) [inline companion/src/burnconfigdialog.cpp:163]: Needs Qt 5.14 - change not supported

File: companion/src/comparedialog.cpp [hunk 6]
```diff
@@ -155,7 +155,7 @@ void CompareDialog::removeModelBtnClicked()
 void CompareDialog::on_printButton_clicked()
 {
   QPrinter printer;
-  printer.setPageMargins(10.0, 10.0, 10.0, 10.0, printer.Millimeter);
+  printer.setPageMargins({10.0, 10.0, 10.0, 10.0}, QPageLayout::Millimeter);
```
Comment (elecpower) [inline companion/src/comparedialog.cpp:158]: Support

File: companion/src/comparedialog.cpp [hunk 7]
```diff
@@ -167,9 +167,9 @@ void CompareDialog::on_printFileButton_clicked()
 {
   QPrinter printer;
   QString filename = QFileDialog::getSaveFileName(this, tr("Select PDF output file"), QString(), "Pdf File(*.pdf)");
-  printer.setPageMargins(10.0, 10.0, 10.0, 10.0, printer.Millimeter);
+  printer.setPageMargins({10.0, 10.0, 10.0, 10.0}, QPageLayout::Millimeter);
```
Comment (elecpower) [inline companion/src/comparedialog.cpp:170]: Support

File: companion/src/comparedialog.cpp [hunk 8]
```diff
@@ -167,9 +167,9 @@ void CompareDialog::on_printFileButton_clicked()
...
   printer.setOutputFormat(QPrinter::PdfFormat);
-  printer.setOrientation(QPrinter::Landscape);
+  printer.setPageOrientation(QPageLayout::Landscape);
```
Comment (elecpower) [inline companion/src/comparedialog.cpp:172]: Support

File: companion/src/firmwares/eeprominterface.cpp [hunk 9]
```diff
@@ -141,7 +141,7 @@ unsigned int Firmware::getVariantNumber()
 {
   unsigned int result = 0;
   const Firmware * base = getFirmwareBase();
-  QStringList options = id.mid(base->getId().length()+1).split("-", QString::SkipEmptyParts);
+  QStringList options = id.mid(base->getId().length()+1).split("-", Qt::SkipEmptyParts);
```
Comment (elecpower) [inline companion/src/firmwares/eeprominterface.cpp:144]: Needs Qt 5.14 - change not supported

File: companion/src/firmwares/opentx/opentxinterface.cpp (hunk 1 above)
Comment (elecpower) [reply companion/src/firmwares/opentx/opentxinterface.cpp:181]: Support - simplest change

File: companion/src/dialogs/filesyncdialog.cpp [hunk 10]
```diff
@@ -104,7 +104,7 @@ static QWidget * folderSelectorWidget(QLineEdit * le, QWidget * parent)
   l->addWidget(btn);
 
   QObject::connect(btn, &QToolButton::clicked, [=]() {
-    QString dir = QFileDialog::getExistingDirectory(parent, fsw->property("fileDialogTitle").toString(), le->text(), 0);
+    QString dir = QFileDialog::getExistingDirectory(parent, fsw->property("fileDialogTitle").toString(), le->text(), {});
```
Comment (elecpower) [inline companion/src/dialogs/filesyncdialog.cpp:107]: Support

File: companion/src/firmwares/opentx/opentxinterface.cpp [hunk 11]
```diff
@@ -178,14 +178,14 @@ bool OpenTxEepromInterface::loadFromByteArray(T & dest, const QByteArray & data)
...
     }
     else {
-      qDebug() << QString().sprintf("%s: Wrong fourcc %x vs %x", getName(), fourcc, Boards::getFourCC(board));
+      qDebug() << QString().asprintf("%s: Wrong fourcc %x vs %x", getName(), fourcc, Boards::getFourCC(board));
```
Comment (elecpower) [inline companion/src/firmwares/opentx/opentxinterface.cpp:184]: Support

File: companion/src/firmwares/opentx/opentxinterface.cpp [hunk 12]
```diff
@@ -178,14 +178,14 @@ bool OpenTxEepromInterface::loadFromByteArray(T & dest, const QByteArray & data)
...
       return false;
     }
   }
-  qDebug() << QString().sprintf("%s: OK", getName());
+  qDebug() << QString().asprintf("%s: OK", getName());
```
Comment (elecpower) [inline companion/src/firmwares/opentx/opentxinterface.cpp:188]: Support

File: companion/src/helpers.cpp [hunk 13]
```diff
@@ -263,7 +263,7 @@ void Helpers::populateFileComboBox(QComboBox * b, const QSet<QString> & set, con
 
   bool added = false;
   // Convert set into list and sort it alphabetically case insensitive
-  QStringList list = QStringList::fromSet(set);
+  QStringList list = QList<QString>(set.begin(), set.end());
```
Comment (elecpower) [inline companion/src/helpers.cpp:266]: Needs Qt 5.14 - change not supported

File: companion/src/logsdialog.cpp [hunk 14]
```diff
@@ -638,7 +638,7 @@ void LogsDialog::saveSession()
     // save the filtered records to a new file
     QString newFilename = logFilename;
     newFilename.append(QString("-Session%1.csv").arg(index));
-    QString filename = QFileDialog::getSaveFileName(this, "Save log", newFilename, "CSV files (.csv);", 0, 0); // getting the filename (full path)
+    QString filename = QFileDialog::getSaveFileName(this, "Save log", newFilename, "CSV files (.csv);", 0, {}); // getting the filename (full path)
```
Comment (elecpower) [inline companion/src/logsdialog.cpp:641]: Supported

File: companion/src/mainwindow.cpp [hunk 15]
```diff
@@ -325,7 +325,7 @@ void MainWindow::closeUpdatesWaitDialog()
 
 QString MainWindow::seekCodeString(const QByteArray & qba, const QString & label) const
 {
-  int posLabel = qba.indexOf(label);
+  int posLabel = qba.indexOf(label.toUtf8());
```
Comment (elecpower) [inline companion/src/mainwindow.cpp:328]: Support

File: companion/src/process_sync.cpp [hunk 16]
```diff
@@ -67,10 +67,10 @@ SyncProcess::SyncProcess(const SyncProcess::SyncOptions & options) :
     m_dirFilters &= ~(QDir::AllDirs);
 
   if (!m_options.includeFilter.isEmpty() && m_options.includeFilter != "*")
-    m_dirIteratorFilters = m_options.includeFilter.split(',', QString::SkipEmptyParts);
+    m_dirIteratorFilters = m_options.includeFilter.split(',', Qt::SkipEmptyParts);
```
Comment (elecpower) [inline companion/src/process_sync.cpp:70]: Needs Qt 5.14 - change not supported

File: companion/src/process_sync.cpp [hunk 17]
```diff
@@ -67,10 +67,10 @@ SyncProcess::SyncProcess(const SyncProcess::SyncOptions & options) :
...
 
   if (!m_options.excludeFilter.isEmpty()) {
-    for (const QString & f : m_options.excludeFilter.split(',', QString::SkipEmptyParts))
+    for (const QString & f : m_options.excludeFilter.split(',', Qt::SkipEmptyParts))
```
Comment (elecpower) [inline companion/src/process_sync.cpp:73]: Needs Qt 5.14 - change not supported

File: companion/src/simulation/simulateduiwidget.cpp (hunk 2 above)
Comment (elecpower) [reply companion/src/simulation/simulateduiwidget.cpp:148]: Needs Qt 5.14 - change not supported

File: companion/src/shared/autobitsetcheckbox.cpp [hunk 18]
```diff
@@ -113,9 +113,11 @@ void AutoBitsetCheckBox::setInverted(bool invert)
 void AutoBitsetCheckBox::updateValue()
 {
   if (m_field)
+  {
```
Comment (elecpower) [inline companion/src/shared/autobitsetcheckbox.cpp:116]: Already fixed in OTX and Companion 2.5

File: companion/src/storage/categorized.cpp [hunk 19]
```diff
@@ -63,7 +63,7 @@ bool CategorizedStorageFormat::load(RadioData & radioData)
     }
 
     // determine if we have a model number
-    QStringList parts = line.split(QRegExp("\\s+"), QString::SkipEmptyParts);
+    QStringList parts = line.split(QRegExp("\\s+"), Qt::SkipEmptyParts);
```
Comment (elecpower) [inline companion/src/storage/categorized.cpp:66]: Needs Qt 5.14 - change not supported

File: companion/src/simulation/widgets/radiowidget.cpp [hunk 20]
```diff
@@ -237,6 +237,6 @@ QDebug operator << (QDebug d, const RadioWidget::RadioWidgetState & o)
 {
   QDebugStateSaver saver(d);
   d << "RadioWidget::RadioWidgetState: type=" << o.type << "; index=" << o.index
-    << "; value=" << o.value << "; flags=0x" << hex << o.flags;
+    << "; value=" << o.value << "; flags=0x" << Qt::hex << o.flags;
```
Comment (elecpower) [inline companion/src/simulation/widgets/radiowidget.cpp:240]: Support these scope qualifiers

File: companion/src/modeledit/curves.cpp [hunk 21]
```diff
@@ -179,10 +179,10 @@ CurvesPanel::CurvesPanel(QWidget * parent, ModelData & model, GeneralSettings &
   }
   QSpacerItem * item = new QSpacerItem(1,1, QSizePolicy::Fixed, QSizePolicy::Expanding);
 
-  ui->curvesLayout->addItem(item,limit + 1, 1, 1, 1, 0);
+  ui->curvesLayout->addItem(item, limit + 1, 1, 1, 1);
```
Comment (elecpower) [inline companion/src/modeledit/curves.cpp:182]: Support

**Comment by elecpower** [comment]
@pfeerick finished my review of the Companion code. Where no comment then previous comments applies for same change

**Comment by pfeerick** [comment]
@sfjuocekr Did you want to work on this further. Otherwise, I can add a commit with the reviewed changes in the #555 PR so that it is included as part of that (if @elecpower is ok with that). Your build script would go in https://github.com/EdgeTX/playground, a new repo to host more experimental/unsupported stuff, and will be the first addition.

**Comment by elecpower** [comment]
@pfeerick @sfjuocekr this PR will need to be rebased on #555 and conflicts resolved.
I have a couple more commits for 555 to push after a bit more testing.

**Comment by sfjuocekr** [comment]
I've had to deal with some unfortunate events in my personal life, sorry for not responding.

The resolve conflicts button is grayed out for me, I'll try to go over the changes and touch up the copy paste errors. No guarantees tho, I've just lost one of my best friends :(

**Comment by pfeerick** [comment]
My condolences. 😢  Take your time... 

Yeah, the changes to `setup_buildenv_ubuntu20.04.sh` were probably too substantial for it to want to let you resolve online. You'll be able to resolve that by removing it - that can get merged into the [playground ](https://github.com/EdgeTX/playground) repo instead.

**Comment by raphaelcoeffic** [comment]
By the way, adding real support for Mac M1 seems to force us to update to C++17 and Qt 6.2 (only supported version so far).

**Comment by elecpower** [comment]
> By the way, adding real support for Mac M1 seems to force us to update to C++17 and Qt 6.2 (only supported version so far).

That is a whole lot of pain as significant changes from Qt 5 to 6.

**Comment by raphaelcoeffic** [comment]
> > By the way, adding real support for Mac M1 seems to force us to update to C++17 and Qt 6.2 (only supported version so far).
> 
> That is a whole lot of pain as significant changes from Qt 5 to 6.

I know… but it’s not like we have a choice…

**Comment by pfeerick** [comment]
Well, we, do... who cares about the Mac M1 :-P That's just one variant of one OS supported... I'm not saying this shouldn't happen, but I don't see it even being worth started on until 2.8/9 ... especially given that means dropping support for Windows 7/ 8 & 10 (pre 1809), macOS 10.13 , and linux distros from 18 -> 20 releases.

**Comment by raphaelcoeffic** [comment]
> Well, we, do... who cares about the Mac M1 :-P That's just one variant of one OS supported... I'm not saying this shouldn't happen, but I don't see it even being worth started on until 2.8/9 ... especially given that means dropping support for Windows 7/ 8 & 10 (pre 1809), macOS 10.13 , and linux distros from 18 -> 20 releases.

Whoa! Really? This is crazy! Qt said they might support Apple silicon in 5.15, but it seems to be only for the “commercial” version, whatever that means.

**Comment by elecpower** [comment]
Upgrade to Qt 5.15.2 covered by #3339 and some of the radio changes have been superseded by PRs moving to new ui, etc
